fix(lint): move tslint to eslint

OlivierAlbertini opened the pull request:
## Which problem is this PR solving?

- closes #891 
- [x] Waiting #969 to be merged in order to finalize circleci file and remove tslint related packages

I will check https://github.com/open-telemetry/opentelemetry-js/issues/1083
Because `.prettierrc.json` through gts changes a lot of files (from 175 to 328).
I have an issue with circleci config where it doesn't see the `.prettierrc.json` (but locally yes).
It doesn't fix this comment https://github.com/open-telemetry/opentelemetry-js/pull/892#discussion_r426547520 even if we add `.prettierrc.json` file through gts (gts and prettier have this rule by default)

OlivierAlbertini commented on packages/opentelemetry-plugin-http/package.json:
```diff
@@ -56,8 +56,9 @@
+    "eslint": "^6.8.0",
     "got": "^9.6.0",
-    "gts": "^1.1.0",
+    "gts": "google/gts#master",
```
I will change to the real version once published

dyladan commented:
GTS version released. Can you update this and get it ready to go out?

OlivierAlbertini commented:
> GTS version released. Can you update this and get it ready to go out?

yes, I would like to validate https://github.com/open-telemetry/opentelemetry-js/blob/7214b5c64a4d8c9fd40456360df1faca92ff88ae/eslint.rules.js with you before going further

I tweaked the circleci file in order to show you that everything is fine, I will finalize the circleci once all packages will be migrated.

dyladan commented on packages/opentelemetry-plugin-http/src/utils.ts:
```diff
@@ -122,7 +121,7 @@ export const hasExpectHeader = (options: RequestOptions): boolean => {
 
   const keys = Object.keys(options.headers);
-  return !!keys.find(key => key.toLowerCase() === 'expect');
+  return !!keys.find((key) => key.toLowerCase() === 'expect');
```
arrow functions with only one argument should not require ()

OlivierAlbertini replied:
we have two options:

```
"as-needed" enforces no braces where they can be omitted.
Object properties for variants of the "as-needed" option:

"requireForBlockBody": true modifies the as-needed rule in order to require parens if the function body is in an instructions block (surrounded by braces).
```

ref: https://eslint.org/docs/rules/arrow-parens

I would prefer requireForBlockBody but as-needed is ok.

dyladan replied:
I prefer as-needed. I don't understand the advantage of requireForBlockBody

dyladan commented on eslint.rules.js:
```diff
@@ -0,0 +1,22 @@
+    "prefer-rest-params": "off",
+    "no-empty": "off", // for tests folder
+    "@typescript-eslint/no-empty-function": "off", // for tests folder
+    "@typescript-eslint/no-explicit-any": "off", // for tests folder
```
Is there a way to only disable this rule within the test folder?

OlivierAlbertini replied:
yes, but it will require to maintain another eslint file in the "tests" folder. Is it ok with you ?

dyladan replied:
yes this is fine

OlivierAlbertini replied:
I figured out how to override rules for specific files within the same eslint file

dyladan commented on eslint.rules.js:
```diff
@@ -0,0 +1,22 @@
+    "no-empty": "off", // for tests folder
+    "@typescript-eslint/no-empty-function": "off", // for tests folder
+    "@typescript-eslint/no-explicit-any": "off", // for tests folder
+    "@typescript-eslint/no-unused-vars": "off", // for tests folder
```
Is there a way to only disable this rule within the test folder?

dyladan commented on eslint.rules.js:
```diff
@@ -0,0 +1,22 @@
+module.exports = {
+    "@typescript-eslint/no-this-alias": "off",
+    "prefer-rest-params": "off",
+    "no-empty": "off", // for tests folder
```
Is there a way to only disable this rule within the test folder?

dyladan commented on eslint.rules.js:
```diff
@@ -0,0 +1,22 @@
+    "prettier/prettier": ["error", { "singleQuote": true }],
+    "node/no-deprecated-api": ["warn"],
+    "header/header": [2, "block", [
+      `\n * Copyright ${new Date().getFullYear()}, OpenTelemetry Authors\n *\n * Licensed under the Apache License, Version 2.0 (the "License");\n * you may not use this file except in compliance with the License.\n * You may obtain a copy of the License at\n *\n *      https://www.apache.org/licenses/LICENSE-2.0\n *\n * Unless required by applicable law or agreed to in writing, software\n * dist ...
```
Looks like this forces current year, rather than year of creation like tslint did. Is there a way to update that or no? Just want to avoid having to bump the year on every file every year.

OlivierAlbertini replied:
done!

OlivierAlbertini commented:
I'm starting other packages

dyladan commented:
Wait on that. Other plugins will be moving to the contrib repo. There is already a PR open for that #969

dyladan commented on packages/opentelemetry-api/src/metrics/Metric.ts:
```diff
@@ -101,7 +101,7 @@ export interface Metric<T> {
 export interface MetricUtils {
   /**
-   * Adds the given value to the current value. Values cannot be negative.
+   * Adds the given value to the current value. Values cannot be negative.
```
what are these whitespace changes?

OlivierAlbertini replied:
it was <nbsp> and it's not recommended by eslint...

dyladan replied:
I don't know how to interpret your answer

OlivierAlbertini replied:
sorry `<>` is hidden in github

OlivierAlbertini replied:
ref: https://eslint.org/docs/rules/no-irregular-whitespace

dyladan commented on packages/opentelemetry-base/src/ExportResult.ts:
```diff
@@ -1,5 +1,5 @@
-/*!
- * Copyright 2019, OpenTelemetry Authors
```
Is it really required to update the year on every copyright header?

OlivierAlbertini replied:
no it was before your comment and I change the eslint config in order to not do that on other plugins

OlivierAlbertini replied:
it's because of "!" and it replace the comment by the template

OlivierAlbertini replied:
"!" was required because the old plugin needed it but with eslint it's not needed anymore and we shouldn't have "!" in the header

OlivierAlbertini replied:
I edited the rule.

dyladan commented on packages/opentelemetry-context-zone-peer-dep/src/ZoneContextManager.ts:
```diff
@@ -141,7 +141,12 @@ export class ZoneContextManager implements ContextManager {
     const contextManager = this;
 
-    return function(this: {}, event: string, listener: Func<void>, opts?: any) {
+    return function (
```
Did line length requirement change?

OlivierAlbertini replied:
need to check the gts plugin...

OlivierAlbertini replied:
In both versions (gts@2.0.0 && gts@1.1.2), no max-length were enforced. Default is `80`. Which is correct here because before it was `82`. I think the eslint parser is just better...

dyladan commented on packages/opentelemetry-exporter-collector/src/CollectorExporter.ts:
```diff
@@ -44,7 +44,7 @@ export class CollectorExporter implements SpanExporter {
   readonly hostName: string | undefined;
   readonly attributes?: Attributes;
-  private _isShutdown: boolean = false;
+  private _isShutdown = false;
```
Can we remove this rule? Generally, I don't see a problem with setting explicit types in these cases.

OlivierAlbertini replied:
It refers to https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-inferrable-types.md

dyladan replied:
Yes I know the rule. I was asking if we can disable it or set `ignoreProperties: true`

dyladan commented on packages/opentelemetry-metrics/src/export/ConsoleMetricExporter.ts:
```diff
@@ -36,11 +36,12 @@ export class ConsoleMetricExporter implements MetricExporter {
       console.log(metric.labels);
       switch (metric.descriptor.metricKind) {
-        case MetricKind.COUNTER:
+        case MetricKind.COUNTER: {
```
Not sure how I feel about requiring this to be a block. I have never seen this syntax before even.

OlivierAlbertini replied:
Same for me too, but after reading this (https://eslint.org/docs/rules/no-case-declarations), it makes sense... WDYT ?

dyladan replied:
its fine

dyladan commented on packages/opentelemetry-plugin-grpc/src/grpc.ts:
```diff
@@ -135,15 +134,13 @@ export class GrpcPlugin extends BasePlugin<grpc> {
       plugin._logger.debug('patched gRPC server');
 
       return function register<RequestType, ResponseType>(
-        // tslint:disable-next-line:no-any
```
Would like to enable the no-any rule, even if we have to put ignores in some places. I try to avoid the use of any wherever possible.

OlivierAlbertini replied:
When you lint your code, you have warnings... and it shows lines, files and the rule you don't respect. It's less noisy.

dyladan commented on packages/opentelemetry-plugin-http/test/utils/httpRequest.ts:
```diff
@@ -52,10 +52,10 @@ function get(input: any, options?: any): GetResult {
-      options != null
-        ? http.get(input, options, onGetResponseCb)
-        : http.get(input, onGetResponseCb);
+    const isValid = options !== null && options !== undefined;
```
What rule triggered this rewrite?

OlivierAlbertini replied:
!= null, it wants !== and it's fairly rare that we use != or ==

dyladan replied:
`!=` can be useful as it matches both null and undefined. Maybe we should allow that?

OlivierAlbertini replied:
I edited the rule.

dyladan commented on packages/opentelemetry-resources/src/platform/node/detectors/EnvDetector.ts:
```diff
@@ -87,10 +87,7 @@ class EnvDetector implements Detector {
-        .trim()
-        .split('^"|"$')
-        .join('');
+      value = value.trim().split('^"|"$').join('');
```
What rule triggered this rewrite?

OlivierAlbertini replied:
I believe that there is no rewrite, I believe that it's just "prettier" that makes one line.

dyladan commented on packages/opentelemetry-test-utils/testUtils.ts:
```diff
@@ -15,8 +15,9 @@
 import * as childProcess from 'child_process';
-import { SpanKind, Attributes, Event, Span, Status } from '@opentelemetry/api';
 import * as assert from 'assert';
+/* eslint-disable node/no-unpublished-import */
```
Why is this an "unpublished import"

OlivierAlbertini replied:
I don't know, perhaps, I googled a bit but nothing clear... Perhaps the package is private... But not sure.

dyladan commented:
All of the force pushing and changing history on this has made it really hard to follow for me, but I don't see any issue right now.

mayurkale22 commented on packages/opentelemetry-api/.eslintignore:
```diff
@@ -0,0 +1,2 @@
+build
```
`build` and `build/` is the same?

mayurkale22 replied:
Maybe, we should include `**/node_modules` directory here?

OlivierAlbertini replied:
yes build and build/ is the same.
We could add `node_modules` but it's ignored by default.

mayurkale22 commented on packages/opentelemetry-api/.eslintignore:
```diff
@@ -0,0 +1,2 @@
+build
+index.ts
```
Why we are ignoring `index.ts`?

OlivierAlbertini replied:
since the PR, we added some rules and indeed, it's not required anymore

mayurkale22 commented on packages/opentelemetry-api/package.json:
```diff
@@ -11,8 +11,8 @@
     "build": "npm run compile",
-    "lint": "gts check",
-    "lint:fix": "gts fix",
+    "lint": "eslint . --ext .ts",
```
I am little confuse about this change, AFAIK gts switched from using [tslint](https://palantir.github.io/tslint/) to [eslint](https://eslint.org/) internally. Looks like here we are changing `gts` to `eslint`. Does it mean we dont need gts now?

OlivierAlbertini replied:
we use it ref: https://github.com/google/gts#working-with-eslint 
https://github.com/open-telemetry/opentelemetry-js/pull/892/files#diff-6759c3472b4d1706bb3ffa02871d49f1R12

dyladan replied:
I actually prefer this way because my eslint plugin for vs code works much better

mayurkale22 commented on packages/opentelemetry-context-base/.eslintignore:
```diff
@@ -0,0 +1,2 @@
+build
```
Is it possible to add this file in top level packages instead within each package?

OlivierAlbertini replied:
No, it's not possible it didn't find the .eslintignore file at the top level

mayurkale22 commented on packages/opentelemetry-context-base/.eslintrc.js:
```diff
@@ -0,0 +1,18 @@
+module.exports = {
```
Same here? It would be simpler to maintain one common file instead of having same boilerplate code in each package.

OlivierAlbertini replied:
It's not always the same since some packages are shared for the web and need some changes.

OlivierAlbertini replied:
I can check if we can centralize a bit more...

mayurkale22 commented on packages/opentelemetry-exporter-collector/src/types.ts:
```diff
@@ -30,7 +31,7 @@ export namespace opentelemetryProto {
         resourceSpans: opentelemetryProto.trace.v1.ResourceSpans[];
       }
-
+      // eslint-disable-next-line @typescript-eslint/no-empty-interface
```
@obecny Do we really need this(`ExportTraceServiceResponse`) empty interface?

obecny replied:
remove it this is not needed

obecny commented on packages/opentelemetry-context-async-hooks/src/AsyncHooksContextManager.ts:
```diff
@@ -103,7 +101,7 @@ export class AsyncHooksContextManager implements ContextManager {
   private _bindFunction<T extends Function>(target: T, context: Context): T {
     const manager = this;
-    const contextWrapper = function(this: {}, ...args: unknown[]) {
+    const contextWrapper = function (this: {}, ...args: unknown[]) {
```
are we sure we want to have this extra space here ?,

OlivierAlbertini replied:
it's matter of being consistant. Sometime we do, sometime we don't. By default, gts require this rule, so I prefer to let it in order to be neutral. Finally, for what I see extra space seems to be more used overall.

obecny replied:
We have already enforced this previously to be without space, now this is going to look like a change not connected to replacing tslint with eslint. I think we should be consistent - doesn't matter if we use tslint eslint or something different, the rules should remain the same, only the mechanism which checks that should be replaced.

OlivierAlbertini replied:
Do you want to deactivate the rule ?

OlivierAlbertini replied:
notice that it's a gts breaking change

obecny replied:
hmm not sure if I understand correctly. We had rule previously which kept no space, isn't there a similar rule ?  And if you deactivate this rule you mentioned does it mean the both cases will be possible or ?

OlivierAlbertini replied:
I made the bad assumption that prettier was configured through eslint in order to have one configuration file but it isn't the case with gts and I need to have a prettier config file that inherits from gts plugin... So by default we have the default rules. I'm fixing this. (gts doesn't care if there is a space or not)

OlivierAlbertini replied:
gts add extra space too

OlivierAlbertini replied:
and it seems to be worst with the prettierrc.json file through the gts plugin
https://github.com/open-telemetry/opentelemetry-js/issues/1083

dyladan commented:
@OlivierAlbertini what is the current status of this?

OlivierAlbertini commented:
> @OlivierAlbertini what is the current status of this?

Hi sorry for the unusual delay.. In getting some issues during the confinement.

I think I addressed all issues, let me know.

dyladan commented:
Looks like your build is failed
